File: learnLeetCode/Day19/Part1/trap_rain_water_ii.cpp
#include <bits/stdc++.h>
using namespace std;

// LC 407 — Trapping Rain Water II

int trapRainWater(vector<vector<int>>& heightMap){
    int n = heightMap.size();
    if (!n) return 0;
    int m = heightMap[0].size();
    if (!m) return 0;
    vector<vector<char>> vis(n, vector<char>(m, 0));

    using T = tuple<int,int,int>; // (height, x, y)
    priority_queue<T, vector<T>, greater<T>> pq;

    // push borders
    for (int i=0;i<n;++i){
        pq.emplace(heightMap[i][0], i, 0); vis[i][0]=1;
        if (m>1){ pq.emplace(heightMap[i][m-1], i, m-1); vis[i][m-1]=1; }
    }
    for (int j=0;j<m;++j){
        if (!vis[0][j]){ pq.emplace(heightMap[0][j], 0, j); vis[0][j]=1; }
        if (n>1 && !vis[n-1][j]){ pq.emplace(heightMap[n-1][j], n-1, j); vis[n-1][j]=1; }
    }

    long long ans = 0;
    int dx[4]={1,-1,0,0};
    int dy[4]={0,0,1,-1};
    while(!pq.empty()){
        auto [h,x,y] = pq.top(); pq.pop();
        for (int k=0;k<4;++k){
            int nx=x+dx[k], ny=y+dy[k];
            if (nx<0||ny<0||nx>=n||ny>=m||vis[nx][ny]) continue;
            vis[nx][ny]=1;
            int nh=heightMap[nx][ny];
            if (nh < h) ans += (h - nh);
            pq.emplace(max(h, nh), nx, ny);
        }
    }
    return (int)ans;
}

int main(){
    vector<vector<int>> grid = {
        {1,4,3,1,3,2},
        {3,2,1,3,2,4},
        {2,3,3,2,3,1}
    };
    cout << trapRainWater(grid) << "\n"; // 4
    return 0;
}

File: learnLeetCode/Day19/Part1/trap_rain_water_ii.py
# LC 407 — Trapping Rain Water II
# Min-heap BFS from the border
import heapq
from typing import List

def trapRainWater(heightMap: List[List[int]]) -> int:
    n = len(heightMap)
    if n == 0:
        return 0
    m = len(heightMap[0])
    if m == 0:
        return 0
    visited = [[False]*m for _ in range(n)]
    pq = []  # (height, x, y)
    # push borders
    for i in range(n):
        heapq.heappush(pq, (heightMap[i][0], i, 0)); visited[i][0] = True
        if m > 1:
            heapq.heappush(pq, (heightMap[i][m-1], i, m-1)); visited[i][m-1] = True
    for j in range(m):
        if not visited[0][j]:
            heapq.heappush(pq, (heightMap[0][j], 0, j)); visited[0][j] = True
        if n > 1 and not visited[n-1][j]:
            heapq.heappush(pq, (heightMap[n-1][j], n-1, j)); visited[n-1][j] = True
    ans = 0
    DIRS = [(1,0),(-1,0),(0,1),(0,-1)]
    while pq:
        h, x, y = heapq.heappop(pq)
        for dx, dy in DIRS:
            nx, ny = x+dx, y+dy
            if nx < 0 or ny < 0 or nx >= n or ny >= m or visited[nx][ny]:
                continue
            visited[nx][ny] = True
            nh = heightMap[nx][ny]
            if nh < h:
                ans += (h - nh)
            heapq.heappush(pq, (max(h, nh), nx, ny))
    return ans

if __name__ == "__main__":
    grid = [
        [1,4,3,1,3,2],
        [3,2,1,3,2,4],
        [2,3,3,2,3,1]
    ]
    print(trapRainWater(grid))  # 4
